test/spec/functions/merge.js

```diff
@@ -0,0 +1,244 @@
+  function Foo() {}
+
+  /* eslint-disable no-new-wrappers */
+  var values = [
+    new Foo(),
+    new Boolean(),
+    new Date(),
+    Foo,
+    new Number(),
+    new String(),
+    new RegExp()
+  ];
+  /* eslint-enable no-new-wrappers */
+
+  var expected = values.map(function() {
+    return true;
+  });
+
+  var actual = values.map(function(value) {
+    var object = merge({}, {a: value, b: {c: value}});
+    return object.a === value && object.b.c === value;
+  });
+
+  expect(actual).toEqual(expected);
+});
+
+it('should not augment source objects 1', function() {
+  var source1 = {a: [{a: 1}]};
+  var source2 = {a: [{b: 2}]};
+  var actual = merge({}, source1, source2);
+
+  expect(source1.a).toEqual([{a: 1}]);
+  expect(source2.a).toEqual([{b: 2}]);
+  expect(actual.a).toEqual([{a: 1, b: 2}]);
+});
+
+it('should not augment source objects 2', function() {
+  var source1 = {a: [[1, 2, 3]]};
+  var source2 = {a: [[3, 4]]};
+  var actual = merge({}, source1, source2);
+
+  expect(source1.a).toEqual([[1, 2, 3]]);
+  expect(source2.a).toEqual([[3, 4]]);
+  expect(actual.a).toEqual([[3, 4, 3]]);
+});
+
+it('should merge plain objects onto non-plain objects', function() {
+  function Foo(object) {
+    Object.assign(this, object);
+  }
+
+  var object = {a: 1};
+  var actual = merge(new Foo(), object);
+
+  expect(actual instanceof Foo).toBe(true);
+  expect(actual).toEqual(new Foo(object));
+
+  actual = merge([new Foo()], [object]);
+  expect(actual[0] instanceof Foo).toBe(true);
+  expect(actual).toEqual([new Foo(object)]);
+});
+
+it('should not overwrite existing values with `undefined` values of object sources', function() {
+  var actual = merge({a: 1}, {a: undefined, b: undefined});
+  expect(actual).toEqual({a: 1, b: undefined});
+});
+
+it('should not overwrite existing values with `undefined` values of array sources', function() {
+  var array = [1];
+  array[2] = 3;
+
+  var actual = merge([4, 5, 6], array);
+  var expected = [1, 5, 3];
+
+  expect(actual).toEqual(expected);
+
+  array = [1, undefined, 3];
+
+  actual = merge([4, 5, 6], array);
+  expect(actual).toEqual(expected);
+});
+
+it('should skip merging when `object` and `source` are the same value', function() {
+  var object = {};
+  var pass = true;
+
+  Object.defineProperty(object, 'a', {
+    configurable: true,
+    enumerable: true,
+    get: function() {
+      pass = false;
+    },
+    set: function() {
+      pass = false;
+    }
+  });
+
+  merge(object, object);
+  expect(pass).toBe(true);
+});
+
+// NOT: we don't want this behavior
+it.skip('should convert values to arrays when merging arrays of `source`', function() {
+  var object = {a: {'1': 'y', 'b': 'z', 'length': 2}};
+  var actual = merge(object, {a: ['x']});
+
+  expect(actual).toEqual({a: ['x', 'y']});
+
+  actual = merge({a: {}}, {a: []});
+  expect(actual).toEqual({a: []});
+});
+
+it('should not convert strings to arrays when merging arrays of `source`', function() {
```

Comment:
We probably don’t have to test this case it’s obvious that we don’t support this kind of operation.

Reply:
likely still  useful, since ES2015 returns keys of strings as if they are arrays, and we definitely don't want to merge strings (it will throw errors)